File: 프로그래머스/3/132266. 부대복귀/부대복귀.java
import java.util.*;

class Solution {
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        
        ArrayList<ArrayList<Integer>> roadInfo = new ArrayList<>();
        
        for(int i=0; i<=n; i++){
            roadInfo.add(new ArrayList<>());
        }
        
        for(int i=0; i<roads.length; i++){
            int from = roads[i][0];
            int to = roads[i][1];
            roadInfo.get(from).add(to);
            roadInfo.get(to).add(from);
        }
        
        int min[] = new int[n+1];
        Arrays.fill(min, Integer.MAX_VALUE);
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        min[destination] = 0;
        pq.add(new int[]{destination, 0});
        
        while(!pq.isEmpty()){
            int now[] = pq.poll();
            if(now[1] > min[now[0]]) continue;
            for(int next : roadInfo.get(now[0])){
                if(now[1]+1 <= min[next]){
                    pq.add(new int[]{next, now[1]+1});
                    min[next] = now[1]+1;
                }
            }
        }
        
        int answer[] = new int[sources.length];
        
        for(int i=0; i<answer.length; i++){
            if(min[sources[i]] == Integer.MAX_VALUE){
                answer[i] = -1;
            }else{
                answer[i] = min[sources[i]];
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/3/132266. 부대복귀/부대복귀.js
function solution(n, roads, sources, destination) {
    
    const roadInfo = {};
    
    for(let i=0; i<roads.length; i++){
        var from = roads[i][0];
        var to = roads[i][1];
        if(!roadInfo[from]) roadInfo[from] = [];
        if(!roadInfo[to]) roadInfo[to] = [];
        roadInfo[from].push(to);
        roadInfo[to].push(from);
    }
    
    const min = Array(n+1).fill(Number.MAX_SAFE_INTEGER);
    const pq = [];
    
    min[destination] = 0;
    pq.push([destination, 0]);
    let head = 0;
    
    while(head < pq.length){
      
        var now = pq[head++];
        if(min[now[0]] < now[1]) continue;
        for(let next of roadInfo[now[0]]){
            if(now[1] + 1 <= min[next]){
                pq.push([next, now[1]+1]);
                min[next] = now[1]+1;
            }
        }
    }
    
    var answer = Array(sources.length).fill(0);
    
    for(let i=0; i<answer.length; i++){
        if(min[sources[i]] == Number.MAX_SAFE_INTEGER){
            answer[i] = -1;
        }else{
            answer[i] = min[sources[i]];
        }
    }
    
    return answer;
}